x-pack/plugins/security/public/session/session_timeout.ts

```diff
@@ -0,0 +1,342 @@
+ */
+export const EXTENSION_THROTTLE_MS = 60 * 1000;
+
+/**
+ * Route to get session info and extend session expiration
+ */
+export const SESSION_ROUTE = '/internal/security/session';
+
+export interface SessionState extends Pick<SessionInfo, 'expiresInMs' | 'canBeExtended'> {
+  lastExtensionTime: number;
+}
+
+export class SessionTimeout {
+  private channel?: BroadcastChannel<SessionState>;
+
+  private isVisible = document.visibilityState !== 'hidden';
+  private isExtending = false;
+
+  private sessionState$ = new BehaviorSubject<SessionState>({
+    lastExtensionTime: 0,
+    expiresInMs: null,
+    canBeExtended: false,
+  });
+  private subscription?: Subscription;
+
+  private warningToast?: Toast;
+
+  private stopActivityMonitor?: Function;
+  private stopVisibilityMonitor?: Function;
+  private removeHttpInterceptor?: Function;
+
+  private stopRefreshTimer?: Function;
+  private stopWarningTimer?: Function;
+  private stopLogoutTimer?: Function;
+
+  constructor(
+    private notifications: NotificationsSetup,
+    private sessionExpired: ISessionExpired,
+    private http: HttpSetup,
+    private tenant: string
+  ) {}
+
+  public start() {
+    if (this.http.anonymousPaths.isAnonymous(window.location.pathname)) {
+      return;
+    }
+
+    this.subscription = this.sessionState$
+      .pipe(skip(1), throttleTime(1000), tap(this.toggleEventHandlers))
+      .subscribe(this.resetTimers);
+
+    return this.fetchSessionInfo();
+  }
+
+  public stop() {
+    const nextState = {
+      lastExtensionTime: 0,
+      expiresInMs: null,
+      canBeExtended: false,
+    };
+    this.toggleEventHandlers(nextState);
+    this.resetTimers(nextState);
+    this.subscription?.unsubscribe();
+  }
+
+  /**
+   * Event handler that receives session information from other browser tabs.
+   */
+  private handleChannelMessage = (message: SessionState) => {
+    this.sessionState$.next(message);
+  };
+
+  /**
+   * HTTP request interceptor which ensures that API calls extend the session only if tab is
+   * visible.
+   */
+  private handleHttpRequest = (fetchOptions: HttpFetchOptionsWithPath) => {
+    // Ignore requests to external URLs
+    if (fetchOptions.path.indexOf('://') !== -1) {
+      return;
+    }
+
+    if (fetchOptions.asSystemRequest === undefined) {
+      return { asSystemRequest: !this.isVisible };
+    }
+  };
+
+  /**
+   * HTTP response interceptor which allows us to prevent extending the session manually if it has
+   * already been extended as part of an API call.
+   */
+  private handleHttpResponse = (httpResponse: HttpResponse) => {
+    // Ignore session endpoint which is already handled by fetch callback
+    if (httpResponse.fetchOptions.path === SESSION_ROUTE) {
+      return;
+    }
+
+    // Extend session unless we're dealing with a system request
+    if (httpResponse.fetchOptions.asSystemRequest === false) {
+      if (this.shouldExtend()) {
+        this.fetchSessionInfo(true);
+      }
+    }
```

Comment:
I know that `this.shouldExtend()` will prevent us from extending the session more than once a minute, but I don't think we even need to do this much. If a non-system-request went through, it already extended the user's session. We check session expiration right before showing the warning, so the only time we need to know about this response is _after_ the warning is visible to the user, right? (because then the warning is no longer valid, their session has already been extended)

Correct me if I'm wrong, but it looks like the way this is currently written: If the warning is NOT showing, and we detected a response to a non-system-request, make an API call to extend the session and fetch new session info.

If so, it seems like we are making extra API calls without any actual change in behavior. The session was already extended, and we already check the latest session info before showing the warning.

Does that sound right to you?

If you agree, also taking into account (comment), I think we should probably change this to:

```ts
if (!httpResponse.fetchOptions.asSystemRequest && this.warningToast) {
  this.fetchSessionInfo();
```

and we could update the `fetchSessionInfo` function so that calling it multiple times before session info has been fetched won't result in multiple HTTP requests.

Reply:
I originally built this so that we mark the session as extended without making an additional API call. Are you proposing to revert this again? 

Ignoring the implementation, which of these behaviours do you want to change / add to?

- extends session when detecting user activity
- extends session when receiving HTTP response
- refetches session info before warning is displayed
- does not extend session if recently extended
- marks HTTP requests as system requests when tab is not visible
- does not mark HTTP requests to external URLs as system requests
- marks HTTP requests as user requests when tab is visible
- resets timers when receiving message from other tabs
- shows warning before session expires
- hides warning if session gets extended
- logs user out slightly before session expires
- logs user out immediately if session expiration is imminent
- does not log user out if session does not expire

Reply:
> * extends session when receiving HTTP response

I don't understand why this has been added, and I don't see the reason for it. It seems like it's adding extra complexity and API calls for no impact on the user experience.

Reply:
The reason this is there is to ensure that the session remains active for e.g. dashboards where no one is actively interacting with the page but the app is polling for changes.

Reply:
But the app polling for changes without `asSystemRequest: true` is already extending the session simply by making the requests. So what is this adding?